mac-virtualcam: Add virtual cam output to OBS as bundled plugin on macOS

PatTheMav opened the pull request:
### Description
Adds the virtual cam plugin written by @johnboiles to OBS main codebase to allow functionality to be shipped with OBS.

To that purpose additional functionality was added to copy the necessary DAL plugin upon first activation of the the virtual cam output.

Minor code changes were also necessary (mainly replacing `typeof` with `__typeof`).

Necessary steps to codesign the DAL plugin and collect dependencies of the plugin itself have been added to CI and macOS build script.

*Notes:*

* ~This code only checks for the existence of the DAL plugin but doesn't check for versioning. Depending on how many changes we expect to this DAL plugin we might need to introduce versioning or a nonce that is valid after a base OBS update (to ensure the plugin is overwritten with every OBS update).~ Fixed via https://github.com/obsproject/obs-studio/pull/3492/commits/a409a359901fd83e696dde4a04e57f50003c7918
* ~When OBS is run from the `dmg` image (without being copied somewhere else on disk) the DAL installation fails right now (probably due to the source path not being correct, I'm investigating this edge case right now).~ Fixed via https://github.com/obsproject/obs-studio/pull/3492/commits/90942718bbf5d51b4f1b6c067da1ee4e274d2dad.

* Once @jp9000 gives his blessing I'll combine the commits/rebase for easier merging

### Motivation and Context
Enables the much-desired feature for macOS users by default.

### How Has This Been Tested?
* DAL plugin Installation tested with bundled OBS as well as run from CLI
* virtual cam output activated with and without DAL plugin installed
* Checked output via Quicktime
* Ran bundling and notarisation
* Checked code-signing on DAL plugin
* Checked for successful notarisation

### Types of changes
- New feature (non-breaking change which adds functionality)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

dodgepong commented:
A couple things:

> To that purpose additional functionality was added to copy the necessary DAL plugin upon first activation of the the virtual cam output.

Not to delay this for too long, but how does this work in terms of UX? Does it ask for elevation when "Start Virtual Camera" is clicked? Ideally we would have an onboarding flow that would handle this on first run of OBS, I would think.

Also, we need to change the "test card" to match the image that we're using for the Windows version.

PatTheMav commented:
> Not to delay this for too long, but how does this work in terms of UX? Does it ask for elevation when "Start Virtual Camera" is clicked? Ideally we would have an onboarding flow that would handle this on first run of OBS, I would think.

Yeah that's how it's solved because I wanted this to be "out" asap. I also don't know how quick we can have the onboarding work.

Nevertheless we need this to work even if a user skips/dismisses the onboarding by accident.

PatTheMav commented:
As for the test card, the macOS one is generated dynamically by the DAL plugin, I'd have to check how to "dumb down" the code to only show the PNG. Mainly @johnboiles can help here?

johnboiles commented:
Thank you for doing this @PatTheMav! I'll take a look today! It's been on my todo list for way too long so I'm glad you got this moving!

Would you mind if I reworked the commits a bit so that my authorship still gets retained in the git history? Probably I'll subtree in my code as the first commit, then apply your other commits on top of it.

As for the image, it will be easy to swap to a static image, but I'm not sure about how to make it modifiable. It may be as simple as having OBS override the png in the plugin, but I'm wondering if that breaks signing. I'll give it a try.

PatTheMav commented:
@johnboiles yeah knock yourself out, it's still your project.

I have to admit I haven't given it full thought how to integrate this with your ongoing efforts (as it would mean you'd have to maintain your plugin as part of an OBS fork), but I happily follow your lead.

johnboiles commented:
Once this is all merged, I'll probably archive my repo and move any ongoing development directly in the obsproject/obs-studio repo. I'm not too attached to the animated test card, and everything else I think we're transitioning over.

johnboiles commented on plugins/mac-virtualcam/src/dal-plugin/Defines.h:
```diff
@@ -0,0 +1,21 @@
+//  You should have received a copy of the GNU General Public License
+//  along with obs-mac-virtualcam. If not, see <http://www.gnu.org/licenses/>.
+
+#define PLUGIN_NAME @"mac-virtualcam"
```
Any strong reason you moved to a static `Defines.h` file instead of sticking with the generated one? Not a huge deal either way, but kinda nice to not have to update the version number in as many places.

But as I write that, updating a few extra lines is probably worth reducing complexity of the CMake build. So I guess I don't have a strong feeling.

PatTheMav replied:
Only reason was the idea that its version would be OBS' version (as it's bundled and all) and I just wanted it to compile.. Was actually thinking about removing the Defines import altogether for that reason. 😄

johnboiles replied:
Oh yeah it's a better idea to just use OBS's version for this string. I assume that is already defined in the build somewhere -- maybe I can use that instead of this code and to generate the .plist file.

johnboiles commented on plugins/mac-virtualcam/src/dal-plugin/Info.plist:
```diff
@@ -0,0 +1,44 @@
+	<key>CFBundleExecutable</key>
+	<string>obs-mac-virtualcam</string>
+	<key>CFBundleIdentifier</key>
+	<string>com.johnboiles.obs-mac-virtualcam.dal-plugin</string>
```
We _might_ want to change this to something more OBS-y

johnboiles commented on plugins/mac-virtualcam/src/obs-plugin/CMakeLists.txt:
```diff
@@ -0,0 +1,59 @@
+    obs-frontend-api
+    Qt5::Core
+    Qt5::Widgets
+	${AVFOUNDATION}
```
I switched all whitespace in the CMake files to tabs so we won't have this inconsistency after merge.

johnboiles commented on plugins/mac-virtualcam/src/obs-plugin/Defines.h:
```diff
@@ -0,0 +1,24 @@
+//  along with obs-mac-virtualcam. If not, see <http://www.gnu.org/licenses/>.
+
+#define PLUGIN_NAME "mac-virtualcam"
+#define PLUGIN_VERSION "0.0.1"
```
I'm changing this to 1.3.0 in my branch to help us debug/understand any sort of issues between folks running the legacy version of my plugin (<=1.2.0) vs folks running the new official OBS version

johnboiles commented on plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm:
```diff
@@ -0,0 +1,201 @@
+	NSString *sourcePath;
+	NSRunningApplication *app = [NSRunningApplication currentApplication];
+
+	if ([app bundleIdentifier] != nil) {
```
Great that you handle both cases!

johnboiles commented on plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm:
```diff
@@ -0,0 +1,201 @@
+			     errorMessage);
+		}
+	} else {
+		blog(LOG_INFO,
```
Idea: I wonder if we should detect this up front and not show the menu item at all. Though if we suspect this will normally work, it's probably not worth worrying about.

PatTheMav replied:
Yeah there might be more elegant variants for this, but my train of thought was this:

* How to get this into the codebase quickest to minimise the gap of time between shipping the feature on Windows and macOS?
* How long until we get the permission onboarding flow (which would include the button to trigger this installation)?
* How to re-enable this functionality if users dismiss the onboarding screen/forget to install the DAL plugin?
* How not to contribute to the amount of prompts happening at first OBS launch already?

I've found that installing it on demand for now using the menu item was quickest and easiest. If we hide the menu item otherwise, we'd need to find a home for the installation button, which would also introduce a dependency to the main UI codebase.

Hence my idea to solve it like this for now and handle it more gracefully as part of the global macOS permission solution later.

johnboiles replied:
Yeah thank you for just knocking it out! Reflecting on it a bit, I think I had been procrastinating because I wasn't sure how to build the permission onboarding flow. I think your approach to get something we can ship and then improve it in the future is right.

This solution is fine by me. I'll leave it to @jp9000 and @dodgepong to determine if it's shippable.

johnboiles commented on plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm:
```diff
@@ -0,0 +1,201 @@
+			@"/Library/CoreMediaIO/Plug-Ins/DAL/obs-mac-virtualcam.plugin";
+		NSString *copyCmd = [NSString
+			stringWithFormat:
+				@"do shell script \"cp -R '%@' '%@'\" with administrator privileges",
```
This is neat

PatTheMav replied:
Yeah originally I tried popping up a directory browsing prompt to get write permissions for the Library folder, but that doesn't work if OBS doesn't run as a sandboxed app and requires to run as an App bundle to pick up the necessary entitlements and all. In the end running an AppleScript was a bit dirty, but did the trick.

johnboiles reviewed (commented):
I reworked the Git history in [johnboiles/macos-virtualcam](https://github.com/obsproject/obs-studio/compare/master...johnboiles:macos-virtualcam?expand=1) to retain authorship. I kept most things the same, but made the following minor changes:

* Removed the PaintCode file for the test card (since we want to get rid of that anyways)
* Changed CMakeLists.txt whitespace from spaces to tabs
* Changed version to 1.3.0 so we can more easily tell the difference between versions of my old plugin (<= 1.2.0) from this new plugin in Console.app logs
* `chmod 644` for source files (they are `chmod 755`'d in this branch)

Probably easiest to just keep this PR open and force push to your branch. Something like this:

```bash
git remote add johnboiles https://github.com/johnboiles/obs-studio.git
git remote add PatTheMav https://github.com/PatTheMav/obs-studio.git
git fetch johnboiles
git checkout johnboiles/macos-virtualcam
git push -f PatTheMav macos-virtualcam
```

Alternately if you add me as a collaborator to your obs-studio fork, I could do this for you.

In case you're interested, this is what I did to maintain history:

```bash
# Create a new branch based on obs-studio:master
git checkout -b macos-virtualcam

# Add johnboiles/obs-mac-virtualcam as a remote
git remote add obs-mac-virtualcam git@github.com:johnboiles/obs-mac-virtualcam.git
# Use Git subtree to squash and merge content from johnboiles/obs-mac-virtualcam into the plugins/mac-virtualcam directory
git subtree add --prefix plugins/mac-virtualcam obs-mac-virtualcam  master --squash --message "Merge from github.com/johnboiles/obs-mac-virtualcam"

# Made a bunch of manual changes to remove unnecessary files. Committed those changes.

# Grab the refs for your fork
git remote add PatTheMav https://github.com/PatTheMav/obs-studio.git
git fetch PatTheMav

# Cherry-pick the first commit and fix merge conflicts (this was the most time-consuming step)
git cherry-pick 0fadfdba

# Cherry-pick everything else (no conflicts)
git cherry-pick c86022af
git cherry-pick 4cc35a95
git cherry-pick 308ddd10
git cherry-pick 90942718

# Push!
git push johnboiles macos-virtualcam
```

johnboiles commented:
Regarding the test card. My plugin [formerly just used an image](https://github.com/johnboiles/obs-mac-virtualcam/blob/37940ff36ea3cb3a74a663bbfa801113c6a158ec/CMIOMinimalSample/Stream.mm#L158), so we can probably easily go back to that. Feel free to take a stab at that part if you have time!

The biggest question in my mind is whether we can swap out the file by just replacing it in-place in `/Library/CoreMediaIO/Plug-Ins/DAL/obs-mac-virtualcam.plugin/Contents/`. I _don't think_ images are part of the plugin bundle signature, but they might be.

PatTheMav commented:
> Alternately if you add me as a collaborator to your obs-studio fork, I could do this for you.

Done. 😂

johnboiles commented:
Ok! I brought my commits in here.

PatTheMav commented:
@johnboiles I haven't been able to find out _why_ it happens exactly, but when OBS is shut down with virtual output active, a single memory leak is reported. When running in DEBUG config I saw that the MachServer's `dealloc` method is called as expected, but then there's one last dangling MachServer status output after OBS flushed its output.

Doesn't seem too pressing, but maybe worth checking out.

PatTheMav commented:
Alright, this was a bit more painful then I expected, but here we go:

* Re-based to current 26.0.0 master
* Changed bundle identifiers to OBS variants
* Changed DAL plugin version to use OBS release version

This part is important as it allows the "installation" procedure to check if the DAL plugin currently installed on the system doesn't belong to the running OBS version and replaces the DAL plugin automatically.

* Removed dynamic placeholder image and added default image from Windows plugin (@dodgepong)

I'd very much welcome some dog-fooding by other team members with macOS machines to get this ready.

PatTheMav commented:
FYI as the DAL plugin is only ad-hoc signed and generated on CI, it will probably not be accepted by any app on macOS. Using Cameo to debug DAL plugin works though (@johnboiles mentioned it on his fork): https://github.com/hivesolutions/cameo

Cameo can also be used to check output of the test card as well as OBS output.

gxalpha commented:
For me QuickTime Player does accept the DAL-plugin without further resigning, but only shows the test card. This is probably the signature issue. (Console.app throwing `Sandbox: QuickTime Player(1992) deny(1) mach-lookup com.obsproject.obs-mac-virtualcam.server`)
Works fine in Cameo! 👍 (im using https://github.com/lvsti/cameo, are multiple of them?).

UX isn't windows equivalent yet since the tools menu still has to be used instead of a dedicated button obviously. 

Some properties, such as "Manufacturer" seem like they still need to be changed (I can probably create a PR to your repo regarding this), also "deviceUID" and "modelUID" probably need a better name, same for "elementName" from the camera stream.
<img width="537" alt="image" src="https://user-images.githubusercontent.com/59806498/94841197-f53c5f00-0419-11eb-89ee-391662bbc998.png">
<img width="574" alt="image" src="https://user-images.githubusercontent.com/59806498/94841459-58c68c80-041a-11eb-9777-cf9e958d3e9f.png">

PatTheMav commented:
Yeah some of those are defined in code, not a big issue to fix them (would love to hear what those values are supposed to be though).

@gxalpha As for the test card, did you get the animated test card in Quicktime or the default static one?

Usually Quicktime is fine with the DAL plugin only if it's signed by a valid developer ID or built on the same machine. Doing a `CMake` run and `make mac-dal-plugin` will allow you to just build the dal plugin so you can copy it over manually.

gxalpha commented:
Regarding what the values are supposed to be:
Apple for their physical FaceTime HD Camera seems to generate an deviceUID for every physical camera (which makes sense but I dont know what vcams would use), and for model UID they enter a different ID for every model so I guess for vcams it should at least contain the version number but it doesn't really matter (This is also the reason why they were ).
For manufacturer it probably depends on what you're using elsewhere. The fact that the entire app is signed by Jim makes me think saying "Hugh Bailey", on the other side "The OBS Project" feels better. "John Boiles" isn't *wrong* either, but doesn't really fit (He needs to get into the Contributors file though).
<img width="684" alt="image" src="https://user-images.githubusercontent.com/59806498/94843492-0dfa4400-041d-11eb-9053-dba461ae43af.png">
Regarding the test card: I had the static one (like in the windows version). Which alternatives would you have expected? (I uninstalled the plugin before)

PatTheMav commented:
No that's exactly the test card you are supposed to see. My Quicktime is a bit hosed, so for a time it still showed the old test card as it somehow retained the old DAL plugin in memory and now it just crashes. 🤷🏻‍♂️

gxalpha commented:
Also FYI:
> The biggest question in my mind is whether we can swap out the file by just replacing it in-place in /Library/CoreMediaIO/Plug-Ins/DAL/obs-mac-virtualcam.plugin/Contents/. I don't think images are part of the plugin bundle signature, but they might be.

Yes, it does work. 
<img width="320" alt="image" src="https://user-images.githubusercontent.com/59806498/94845015-5adf1a00-041f-11eb-9998-4a677e7d770b.png">
Other aspect ratio files will get cropped though
<img width="320" alt="image" src="https://user-images.githubusercontent.com/59806498/94845299-c9bc7300-041f-11eb-94ca-9f3d5773bd7f.png">
> so for a time it still showed the old test card as it somehow retained the old DAL plugin in memory and now it just crashes. 🤷🏻‍♂️

Have you tried to completely uninstall, then restart the computer, try QucikTime again, if it works then install the new DAL plugin? Updating the plugin (just from version to version) has been a pain for many people, since the installer for some reason didn't overwrite the DAL plugin. This might be the same issue here.
If the steps above didn't work, just wait, it should magically resolve itself (no one knows why). If it then is your QuickTime player still crashing maybe try reinstalling it, but I didn't find it in the app store so be careful :D
So yeah, you'll have a fun time maintaining this, including the *few* compatibility issues...

PatTheMav commented:
> Yes, it does work

Yeah that's why I built it that way - it was the easiest way and also the most macOS-like.

> Have you tried to completely uninstall, then restart the computer

Yeah that'll probably fix it - I usually don't restart my iMac in months, so I'll take the chance and do some cleanup before a restart. 😂

gxalpha commented:
Another thing I noticed is that since the test card remembers the resolution of the last incoming stream and uses that (because some apps had issues with cameras changes resolution), the test card will also get cropped. This one is kind of obvious if you think about it but I wanted to note it still.
<img width="320" alt="image" src="https://user-images.githubusercontent.com/59806498/94848437-49e4d780-0424-11eb-9444-6e4bf1b9fcb7.png">

PatTheMav commented:
Yeah from what I could tell in code the DAL plugin stores the resolution via `NSUserDefaults` and re-uses those values. I haven't checked if that actually works, but sounds like that's what happens there.

gxalpha commented:
I can confirm that it does work (I'm the one who added it).
It wasn't an issue (or at least was a relatively easy fix) with the original test card since it got drawn from code (as you already know). I wonder if we can notify the user maybe when the test card has a wrong aspect ratio (Though this should be added if the windows version does something similar, and I have no idea what the windows version does :D)

PatTheMav commented:
@gxalpha I just pushed a commit that ensures the DAL plugin keeps the placeholder image's aspect ratio and resizes/centers it accordingly over a black background.

It will still "remember" the last stream's resolution, but won't squash the image now.

gxalpha commented:
I see, will test later this week when back home

PatTheMav commented:
Commit https://github.com/obsproject/obs-studio/pull/3492/commits/04704072d3f77517ae8cf90187f15f810e3af136 implements the built-in Virtual Camera API, so now functionality is enabled and disabled using the main UI button as requested by @DDRBoxman.

PatTheMav commented:
Ping @johnboiles.

This has been tagged for release with OBS v26.1 and I think I got all feature/change requests in from reviews.

In a next step I'd like to clean up files (e.g. removing the now deprecated dynamic testcard code) and squash commits to have cleaner merge with the main OBS codebase. Due to the subtree merge the commits can't all be squashed so I don't know how to proceed here as the history isn't needed for a single-commit merge anyway.

You should get all the credit here, I just worked on the maintenance task of bringing this into the OBS codebase.

gxalpha commented:
Just wanted to note that https://github.com/johnboiles/obs-mac-virtualcam/issues/160 is still a thing, and unless someone with more macOS experience wants to look into it it would still going to be a thing in whatever gets merged here

jp9000 commented:
Yea I'm aware of it. A developer needs to be able to reproduce, seems like it's currently unreliably reproduced at the moment, which makes it difficult to debug.

PatTheMav commented:
So I've tried to replicate the bug the last few days and wasn't able to - tried all kinds of weird resolutions, aspect ratios and mismatches between canvas and output and except for a cropped image (because Chrome keeps the resolution it accessed the device first with) I couldn't get it to behave like in the bug reports.

@jp9000 do you want this squashed into a single commit? I might need to remove @johnboiles' subtree from the commit for this or squash everything after https://github.com/obsproject/obs-studio/pull/3492/commits/781fbe135f303be4c531d75d1ae3a0f17bb54d02 (or I'm mistaken and need to learn a bit more `git`).

PatTheMav commented:
Cleaned-up commits and re-based on master. I've retained the subtree-add for now.

jp9000 commented:
I feel like this may need to be rebased. Not sure if we should try to use subtree.

johnboiles commented:
Yeah if we're going to squash this before merging it will lose the subtree stuff but that's totally fine if that's @jp9000's preference for merging!

jp9000 commented:
Are we using subtree to preserve history? Is the history worth preserving? Especially considering everything probably had to be reformatted with clang-format I'm assuming.

PatTheMav commented:
@jp9000 yeah clang touched on nearly every file I'd guess. I'm fine with squashing everything for the PR and keeping the subtree+history in a separate branch so @johnboiles can fork or add my repos as remote and pull those changes in from there.

johnboiles commented:
Yeah totally fine to blow away my subtree stuff and squash merge. I was hoping to get in the commit/contributor history just for vanity sake, but I can live without it :)

johnboiles commented:
As for the [skewed output](https://github.com/johnboiles/obs-mac-virtualcam/issues/160) bug, my money's on [this section](https://github.com/johnboiles/obs-mac-virtualcam/blob/master/src/dal-plugin/CMSampleBufferUtils.mm#L41-L55) of code. This could previously be reproduced consistently by using certain nonstandard resolutions IIRC so I made [this change](https://github.com/johnboiles/obs-mac-virtualcam/commit/9ed5fafdee934a0f48392243f170129f5eaab96b) to use [`CVPixelBufferGetBytesPerRowOfPlane`](https://developer.apple.com/documentation/corevideo/1456711-cvpixelbuffergetbytesperrowofpla?language=objc) which made things better than they were before (When I would just assume the pixel buffer was just one big block of memory. Though now that I pull up that documentation I noticed [`CVPixelBufferGetBaseAddressOfPlane`](https://developer.apple.com/documentation/corevideo/1456821-cvpixelbuffergetbaseaddressofpla?language=objc). I _bet_ I need to use that instead of `CVPixelBufferGetBaseAddress`. Hopefully that fixes it!

👏 thank you so much @PatTheMav for keeping this moving and thank you @gxalpha for _all_ the work you've put in troubleshooting and fixing things over the last 7 months. I have huge appreciation for the both of you.

PatTheMav commented:
@johnboiles I guess Jim could make an exception and add you regardless of your position in the contributor ranking, that’s his choice though. Otherwise cherry-pick the commits without commits and open a new PR, I’d close/withdraw mine then. Your call.

johnboiles commented:
Nah, it's not worth worrying about. Whatever Jim thinks is best / normally does I'm fine with.

dodgepong commented:
@johnboiles Regarding that potential fix, what is the best way for that to be tested? Can you (or someone) implement that fix and release a test build somewhere?

PatTheMav commented:
Got it - think it’s worth to check out the `CVPixelBuffer` thing first before we merge this? I think you also thought about improving performance by using a memory buffer without copy?

johnboiles commented:
I committed the change via the GitHub web UI, but I'm not able to try it right now. I think it's worth a shot. I strongly suspect that's the _more correct_ API to be using.

johnboiles commented:
I took a quick stab at fixing the johnboiles/obs-mac-virtualcam build in https://github.com/johnboiles/obs-mac-virtualcam/pull/229 and then created a PR on top of that with this `CVPixelBufferGetBytesPerRowOfPlane` change in  https://github.com/johnboiles/obs-mac-virtualcam/pull/230.

Assuming I fixed the build, then there should be a signed pkg in the build artifacts of that second PR we can try out.

gxalpha commented:
> 👏 thank you so much @PatTheMav for keeping this moving and thank you @gxalpha for _all_ the work you've put in troubleshooting and fixing things over the last 7 months. I have huge appreciation for the both of you.

You're welcome :)

> Yeah totally fine to blow away my subtree stuff and squash merge. I was hoping to get in the commit/contributor history just for vanity sake, but I can live without it :)

Well that doesn't mean you won't get into the Contributors document, does it?

johnboiles commented:
IDK! I don't so much care about the document that's checked into source code but it would be nice to be somewhere in the GitHub contributors list or in the commit history! Though maybe that will happen no matter what? But really y'all not worth worrying much about.

> I think you also thought about improving performance by using a memory buffer without copy?

Yes! The code to possibly avoid another memory copy is in [here](https://github.com/johnboiles/obs-mac-virtualcam/blob/d31d1795e0e5acf6bf5fbaf947bd75edcee07428/src/dal-plugin/CMSampleBufferUtils.mm#L102). It mostly worked! It just for some reason that I don't understand would cause the virtualcam source to not work inside of OBS (as a loopback). This was the main way I tested while developing so it was a dealbreaker for me for that to not work. I was also a little nervous it could break some other app, though in my testing, apps other than OBS seemed to work fine. It'd be an interesting one to look into for sure, but I think for now it's probably better to keep the extra memory copy that seems to work more reliably.

Ok build fix worked, so the pkg will be in the artifacts of [this build](https://github.com/johnboiles/obs-mac-virtualcam/runs/1282989351?check_suite_focus=true) once it finished. I'm about to get off the internet for the day, but I maybe can try it tomorrow if no one else can in the meantime.

PatTheMav commented:
Squash is done, including that last fix by John - I've also made him the author of the commit. I hope CI doesn't take a shit again..

gxalpha commented:
Is there something like an uninstaller included? Many people will be heavily confused if they delete OBS but still see the DAL-plugin showing up under cameras.

Also whats with the other contributors to the plugin? Personally I don't mind either (would be cool, but i dont wanna halt progress) but the other 15 might (https://github.com/johnboiles/obs-mac-virtualcam/graphs/contributors).

PatTheMav commented:
There are usually no such things as uninstallers on macOS or Linux - the most common solution I've seen is a shell script provided either with the `dmg` image or (which would suffice in our case) a single command mentioned in the FAQ/wiki to that purpose.

As for the other contributors that's not something I can answer - the practice for PRs is to squash into a single commit and a commit cannot have multiple authors.

Gol-D-Ace commented:
It's still possible to preserve the authors like this:
https://docs.github.com/en/free-pro-team@latest/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors

PatTheMav commented:
Oh Co-Authors, that's nifty, didn't know about that. If someone give me a list of names+emails, I'll amend the commit and force-push it, no problem. Dunno how the stats will parse that, but I guess that's the least of the issues..

gxalpha commented:
I could create such a list (if the emails are public), I'll send it later

johnboiles commented:
Agh my [test build](https://github.com/johnboiles/obs-mac-virtualcam/actions/runs/320355814) did not work in OBS

```
error: os_dlopen(/Library/Application Support/obs-studio/plugins/obs-mac-virtualcam/bin/obs-mac-virtualcam.so->/Library/Application Support/obs-studio/plugins/obs-mac-virtualcam/bin/obs-mac-virtualcam.so): dlopen(/Library/Application Support/obs-studio/plugins/obs-mac-virtualcam/bin/obs-mac-virtualcam.so, 257): Library not loaded: @executable_path/../Frameworks/QtWidgets.framework/Versions/5/QtWidgets
  Referenced from: /Library/Application Support/obs-studio/plugins/obs-mac-virtualcam/bin/obs-mac-virtualcam.so
  Reason: Incompatible library version: obs-mac-virtualcam.so requires version 5.15.0 or later, but QtWidgets provides version 5.14.0
```

I think maybe homebrew is using a newer QT than OBS now. I'll try to pin it to 5.14 and try again.

PatTheMav commented:
Yeah Homebrew should be on 5.15.1 or something, we're still on 5.14.1 - which is why the `QTDIR` switch was added to `cmake` so you can point it to the Qt we provide as part of obs-deps.

gxalpha commented:
I got the following results by searching for "email" at https://api.github.com/users/GITHUB_USERNAME/events/public, which is a public api; and/or by looking at their GitHub profiles.
Results sorted by number of commits according to https://github.com/johnboiles/obs-mac-virtualcam/graphs/contributors

@johnboiles (John Boiles)(94 commits): johnaboiles@gmail.com
@lvsti (44 commits): lvsti@users.noreply.github.com (email private)
@gxalpha (Sebastian Beckmann)(19 commits): beckmann.sebastian@outlook.de
@signalwerk (Stefan Huber)(2 commits): sh@signalwerk.ch
@seanchas116 (Ryohei Ikegami)(2 commits): iofg2100@gmail.com
@colindean (Colin Dean)(2 commits): colin.dean@target.com

(Everyone starting here 1 commit)
@wuerzelchen (Wolfgang Ladermann): extern.ladermann_wolfgang@allianz.de
@simoneves (Simon Eves): simon.eves@omnisci.com
@orthros (Colin Nelson): colnnelson@google.com
@niw (Yoshimasa Niwa): niw@niw.at
@mkarliner (Michael Karliner): mike@modern-industry.com
@leonelsamayoaup (Leonel Samayoa): Can't find out
@jocamero: Can't find out
@jasongrout (Jason Grout): jgrout6@bloomberg.net
@jadenguy (Alfredo Inostroza): jadenguy@gmail.com
@iKenndac (Daniel Kennett): daniel@cascable.se
@gep13 (Pascal Berger): pascal.berger@gmail.com
@cienijr (José Carlos Cieni Júnior): Can't find out

*This information is subject to errors*

gep13 commented:
Thanks for the shout out. The email address attributed to my GitHub user name is incorrect. That should be gep13@gep13.co.uk Thanks

PatTheMav commented:
@gep13 Yeah sorry, missed that in the last force-commit. There's gonna be another last-minute commit anyway which will include the correct mail then I guess 😉.

cienijr commented:
Hi there, my email is cienijr@outlook.com

gxalpha commented:
I guess since everyone got pinged if someone also has corrections to make they should post it here(?) and I'll update my list

johnboiles commented:
Ok I got my build all fixed up and I'll try to [recruit the folks from johnboiles/obs-mac-virtualcam#160](https://github.com/johnboiles/obs-mac-virtualcam/issues/160#issuecomment-713806186) to see if they can repro the issue and test I solved it. 🤞 hopefully we get some good reports!

johnboiles commented:
I realized I need to do the same thing for the test card as well and committed the change. Corresponds to [this commit](https://github.com/johnboiles/obs-mac-virtualcam/commit/4bd585204ae220068bd55eddf7239b9c8fd8b1dc) in johnboiles/obs-mac-virtualcam

PatTheMav commented:
Yeah OBS clang-format rules and Xcode/macOS code conventions don't see each other eye to eye.. 😂

I'm going to force-push that fix (and a few co-author fixes) to clean this up again.

gep13 commented:
@PatTheMav is there a missing `>` after my email address in the commit message?

PatTheMav commented:
@gep13 I have no idea what you are talking about.. 😉

jp9000 commented:
Well, I hope everyone's ready for the big merge.

I just wanted to say that everyone who contributed to the plugin did an excellent job working together, and you've made me proud. That being said, I hope we're ready to manage this much code. I'm going to hope that this doesn't cause us nightmares down the road considering how often Apple loves to break their own APIs. This is a lot of code for me personally to manage on my own, so I need people who have worked on this to help out as much as their able if it needs maintenance in the future. If anything goes wrong in the future, I'm not sure if I'll be able to handle it on my own.

I just needed to state that as a clear disclaimer before I merge it.

johnboiles commented:
We did it! High fives all around!

I've really appreciated all of y'all's collaboration. Special thanks to @PatTheMav for getting this across the finish line and to @gxalpha for being my biggest partner in managing issues support and fixes for my repo. And thank you @jp9000 for your patience and guidance. ❤️❤️❤️

gxalpha commented:
Yay, very nice!

It's been a pleasure 😄

gep13 commented:
@johnboiles @gxalpha it is great to see this getting merged in, well done!

xiangbao commented:
@PatTheMav  @johnboiles I used a mac-virtual camera with Microsoft PowerPoint Recording	but failed to find the server. if (![[weakSelf machClient] isServerAvailable]) { DLog(@"Server is not available");}
But the OBS app works with Microsoft Powerpoint without a problem. Another thing my app can work well like room,loom, and so on.  Could you give me some advice to fix this problem?  Should I do some config in my own app? I have been stuck for two days. Thanks.
